### PR TITLE
do not send unhealthy endpoints in nds for headless

### PR DESCRIPTION
**Pull request by ramaraochavali**
Regression from https://github.com/istio/istio/pull/47379

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Does this bypass `publishNotReadyAddresses` on service?

**Comment by ramaraochavali** [comment]
> Does this bypass publishNotReadyAddresses on service?

Yes it is. I think it was not handled earlier also. But we can handle it now.

**Review by howardjohn — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
should backport

/cherrypick release-1.23

/cherrypick release-1.22

/cherrypick release-1.21

File: pkg/dns/server/name_table.go [hunk 1]
```diff
@@ -70,7 +70,8 @@ func BuildNameTable(cfg Config) *dnsProto.NameTable {
 							break
 						}
 					}
-					if len(instance.Addresses) == 0 || !isValidInstance {
+					if len(instance.Addresses) == 0 || !isValidInstance ||
+						(!svc.Attributes.PublishNotReadyAddresses && instance.HealthStatus == model.UnHealthy) {
```
Comment (hzxuzhonghu) [inline pkg/dns/server/name_table.go:74]: Can you add a unit test

**Pull request merged**